### PR TITLE
hooks: refactor GObject introspection (gi) hooks

### LINKED ISSUE #6897: Pyinstaller 5.x quits freezing when importing gi.pygtkcompat
Hi, I found issue of pyinstaller 5.x quits freezing when 'import gi.pygtkcompat' encountered in a source-file('twisted' in this case). This happens because that module has various gstreamer import references in one of it's functions, and hence stops freezing with error of missing gst typelibs/namespace as not having gstreamer built.

Can be easilly reproduced with 'pyinstaller gst.py', and a 'gst.py' of just 'import gi.pygtkcompat'. BTW, also happens with 'import pygtkcompat'. 

Win10, python 3.9.13(also tested 3.10.5), latest pygobject/gtk3 built from gvsbuild and added to %PATH%.

This was also issue in pyinstaller v4.x, showing the error on the cmd-output but just didn't quit the freeze - I believe you changed pyinstaller v5.x to be more effective against quiting upon child-process errors, so guessing that's why.

Current workaround is with pyinstaller 5.x using an 'excludes' directive in spec of:
```
excludes=['gi.repository.Gst','gi.repository.GstAudio','gi.repository.GstVideo','gi.repository.GstBase','gi.repository.GstPbutils'],
```
Which works, but was wondering if this is something need more looking into from your side, if possible, thanks - or if we just need to add such meassures ourself because of the way this works in general?

Thanks alot in advance!
```
C:\deluge2\deluge-build>pyinstaller gst.py
546 INFO: PyInstaller: 5.1
546 INFO: Python: 3.9.13
562 INFO: Platform: Windows-10-10.0.19044-SP0
562 INFO: wrote C:\deluge2\deluge-build\gst.spec
578 INFO: UPX is not available.
593 INFO: Extending PYTHONPATH with paths
['C:\\deluge2\\deluge-build']
pygame 2.1.2 (SDL 2.0.18, Python 3.9.13)
Hello from the pygame community. https://www.pygame.org/contribute.html
3108 INFO: checking Analysis
3125 INFO: Building because C:\deluge2\python\Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_gstreamer.py changed
3140 INFO: Initializing module dependency graph...
3140 INFO: Caching module graph hooks...
3264 INFO: Analyzing base_library.zip ...
9640 INFO: Processing pre-find module path hook distutils from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-distutils.py'.
9640 INFO: distutils: retargeting to non-venv dir 'C:\\deluge2\\python\\lib'
17296 INFO: Caching module dependency graph...
17640 INFO: running Analysis Analysis-00.toc
17687 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by C:\deluge2\python\python.exe
17861 INFO: Analyzing C:\deluge2\deluge-build\gst.py
18328 INFO: Processing pre-safe import module hook gi.repository.GObject from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GObject.py'.
18343 INFO: Processing pre-safe import module hook gi.repository.GLib from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GLib.py'.
18343 INFO: Processing pre-safe import module hook gi.repository.Gio from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Gio.py'.
18343 INFO: Processing pre-safe import module hook gi.repository.Atk from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Atk.py'.
18359 INFO: Processing pre-safe import module hook gi.repository.Pango from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Pango.py'.
18359 INFO: Processing pre-safe import module hook gi.repository.PangoCairo from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.PangoCairo.py'.
18359 INFO: Processing pre-safe import module hook gi.repository.Gdk from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Gdk.py'.
18359 INFO: Processing pre-safe import module hook gi.repository.GdkPixbuf from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GdkPixbuf.py'.
18359 INFO: Processing pre-safe import module hook gi.repository.Gtk from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Gtk.py'.
18423 INFO: Processing pre-safe import module hook gi.repository.Gst from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.Gst.py'.
18423 INFO: Processing pre-safe import module hook gi.repository.GstAudio from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GstAudio.py'.
18423 INFO: Processing pre-safe import module hook gi.repository.GstVideo from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GstVideo.py'.
18438 INFO: Processing pre-safe import module hook gi.repository.GstBase from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GstBase.py'.
18438 INFO: Processing pre-safe import module hook gi.repository.GstPbutils from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GstPbutils.py'.
18438 INFO: Processing module hooks...
18438 INFO: Loading module hook 'hook-difflib.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
18452 INFO: Loading module hook 'hook-distutils.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
18452 INFO: Loading module hook 'hook-encodings.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
18672 INFO: Loading module hook 'hook-gi.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
18718 INFO: Loading module hook 'hook-gi.repository.Atk.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
20359 INFO: Loading module hook 'hook-gi.repository.Gdk.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
21391 INFO: Processing pre-safe import module hook gi.repository.cairo from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.cairo.py'.
21406 INFO: Loading module hook 'hook-gi.repository.GdkPixbuf.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
24452 INFO: Processing pre-safe import module hook gi.repository.GModule from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.repository.GModule.py'.
24452 INFO: Loading module hook 'hook-gi.repository.Gio.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
26593 INFO: Loading module hook 'hook-gi.repository.GLib.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
29110 INFO: Loading module hook 'hook-gi.repository.GModule.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
30046 INFO: Loading module hook 'hook-gi.repository.GObject.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
31030 WARNING: Hidden import "gi._gobject" not found!
31030 INFO: Loading module hook 'hook-gi.repository.Gst.py' from 'C:\\deluge2\\python\\lib\\site-packages\\PyInstaller\\hooks'...
Traceback (most recent call last):
  File "C:\deluge2\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\deluge2\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\deluge2\python\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
  File "C:\deluge2\python\lib\site-packages\PyInstaller\__main__.py", line 178, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "C:\deluge2\python\lib\site-packages\PyInstaller\__main__.py", line 59, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\building\build_main.py", line 842, in main
    build(specfile, distpath, workpath, clean_build)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\building\build_main.py", line 764, in build
    exec(code, spec_namespace)
  File "C:\deluge2\deluge-build\gst.spec", line 7, in <module>
    a = Analysis(
  File "C:\deluge2\python\lib\site-packages\PyInstaller\building\build_main.py", line 319, in __init__
    self.__postinit__()
  File "C:\deluge2\python\lib\site-packages\PyInstaller\building\datastruct.py", line 173, in __postinit__
    self.assemble()
  File "C:\deluge2\python\lib\site-packages\PyInstaller\building\build_main.py", line 487, in assemble
    self.graph.process_post_graph_hooks(self)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\depend\analysis.py", line 326, in process_post_graph_hooks
    module_hook.post_graph(analysis)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\depend\imphook.py", line 398, in post_graph
    self._load_hook_module()
  File "C:\deluge2\python\lib\site-packages\PyInstaller\depend\imphook.py", line 361, in _load_hook_module
    self._hook_module = importlib_load_source(self.hook_module_name, self.hook_filename)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\compat.py", line 620, in importlib_load_source
    return mod_loader.load_module()
  File "<frozen importlib._bootstrap_external>", line 529, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 1029, in load_module
  File "<frozen importlib._bootstrap_external>", line 854, in load_module
  File "<frozen importlib._bootstrap>", line 274, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 711, in _load
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "C:\deluge2\python\lib\site-packages\PyInstaller\hooks\hook-gi.repository.Gst.py", line 30, in <module>
    binaries, datas, hiddenimports = get_gi_typelibs('Gst', '1.0')
  File "C:\deluge2\python\lib\site-packages\PyInstaller\utils\hooks\gi.py", line 71, in get_gi_typelibs
    typelibs_data = _gi_typelibs(module, version)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\isolated\_parent.py", line 269, in wrapped
    return call(function, *args, **kwargs)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\isolated\_parent.py", line 238, in call
    return isolated.call(function, *args, **kwargs)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\isolated\_parent.py", line 176, in call
    raise RuntimeError(f"Child process call to {function.__name__}() failed with:\n" + output)
RuntimeError: Child process call to _gi_typelibs() failed with:
  File "C:\deluge2\python\lib\site-packages\PyInstaller\isolated\_child.py", line 63, in run_next_command
    output = function(*args, **kwargs)
  File "C:\deluge2\python\lib\site-packages\PyInstaller\utils\hooks\gi.py", line 63, in _gi_typelibs
    repo.require(module, version, GIRepository.RepositoryLoadFlags.IREPOSITORY_LOAD_FLAG_LAZY)
gi.repository.GLib.GError: g-irepository-error-quark: Typelib file for namespace 'Gst', version '1.0' not found (0)
```

### PR DESCRIPTION
**Pull request by rokm**
The modules imported from `gi.repository` are marked as runtime modules by their corresponding pre-safe-import-module hooks. Therefore, their standard hooks are always loaded and executed, regardless of whether the modue is actually importable or not.

In PyInstaller v5, this behavior triggers errors in hooks for GI modules that are not importable, because the new `isolated` framework propagates the errors instead of swallowing them. While these errors could be caught and demoted to warnings
to match the old behavior, it would be better if hooks checked whether module is importable before doing any processing at all.

To that end, we introduce new class, `GiModuleInfo` that, as part of its initialization, allows us to:
 - perform availability check
 - obtain data previously returned by `get_gi_typelibs`
 - obtain data previously returned by `get_gi_libdir`
 
using a single isolated import attempt (instead of one being performed in each of those steps).

In  addition, if passed `hook_api` as an optional argument, the `GiModuleInfo` can use hook configuration API to override the GI module version to be collected (which allows the standard use pattern to be removed from the hook itself).

The old `get_gi_typelibs` and `get_gi_libdir` functions now internally use `GiModuleInfo` to provide backward compatible behavior to (potential) external user.

All `gi` hooks are ported to the `GiModuleInfo` and now become no-op if the module is not available.

In addition, hooks are cleaned up/refactored so that all processing is performed either in the loading stage ("simple" hooks that do not require access to hook configuration API) or in the `hook()` function (hooks that require access to hook configuration API), but not in the mixture of the two.

Fixes #6897.

### CODE REVIEW

File: PyInstaller/hooks/hook-gi.repository.Gtk.py
Comment (zevlee) [inline PyInstaller/hooks/hook-gi.repository.Gtk.py]: Just a little typo on line 47. # Tanslations -> # Translations

Comment (rokm) [reply PyInstaller/hooks/hook-gi.repository.Gtk.py]: Thanks!

File: PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py
Comment (bwoodsend) [inline PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py]: I know that you didn't write this bit but I very much doubt that the suffix will ever not be `.so` on non-Windows.

File: PyInstaller/utils/hooks/gi.py
Comment (bwoodsend) [inline PyInstaller/utils/hooks/gi.py:52]: This would handle the stripping and the `sharedlibs == ""` case slightly more seamlessly.
```suggestion
            sharedlibs = re.findall(r"[^,\s]+", sharedlibs)
```

**Review by bwoodsend — COMMENTED** [review]
Ughh, scary gi stuff I don't understand.

**Comment by bwoodsend** [comment]
@danyeaw I believe that you use gi and like to install PyInstaller straight from `develop` so if this change contains any regressions, it'll probably be your CI that breaks first. Perhaps if you do a trial CI run of your gi+PyInstaller projects but installing PyInstaller from this branch, that might save you the fiasco if figuring out what broke it after we hit the merge button?

File: PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py
Comment (danyeaw) [reply PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py]: Does this mean that we can't bundle GTK apps in Windows with this change?

**Comment by danyeaw** [comment]
Thanks @bwoodsend, although I am not that much of a 🤠  to run PyInstaller on CI right from develop, I would be happy to help test this 👍

File: PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py
Comment (rokm) [reply PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py]: > Does this mean that we can't bundle GTK apps in Windows with this change?

It shouldn't; the `lib_ext` should be set to `*.dll` on Windows, same as it was for its `ext` counterpart before the refactoring. 

Functionally, the difference here is that for unsupported platforms, we now display a warning and return an empty list, whereas previously we would end up with exception due to `ext` not being set...

Comment (danyeaw) [reply PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py]: Ok, great!

Comment (rokm) [reply PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py]: OK, we now assume `*.so` pattern on all non-Windows, which I suppose is reasonable enough. And probably doesn't make things much worse, anyway.

File: PyInstaller/utils/hooks/gi.py
Comment (rokm) [reply PyInstaller/utils/hooks/gi.py:52]: We still need the `if sharedlibs else []` part, because we must handle the `sharedlibs` being  `None`.

I also think split + strip makes it more obvious what is being done here, although that might just be my relative lack of exposure to regexes :)

**Comment by danyeaw** [comment]
I am getting a SyntaxError:

```
6348 INFO: Loading module hook 'hook-gi.repository.GdkPixbuf.py' from '/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/hooks'...
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/.venv/bin/pyinstaller", line 8, in <module>
    sys.exit(run())
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/__main__.py", line 178, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/__main__.py", line 59, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 842, in main
    build(specfile, distpath, workpath, clean_build)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 764, in build
    exec(code, spec_namespace)
  File "gaphor.spec", line 24, in <module>
    a = Analysis(
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 319, in __init__
    self.__postinit__()
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/datastruct.py", line 173, in __postinit__
    self.assemble()
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 487, in assemble
    self.graph.process_post_graph_hooks(self)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/analysis.py", line 326, in process_post_graph_hooks
    module_hook.post_graph(analysis)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/imphook.py", line 398, in post_graph
    self._load_hook_module()
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/imphook.py", line 361, in _load_hook_module
    self._hook_module = importlib_load_source(self.hook_module_name, self.hook_filename)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/compat.py", line 620, in importlib_load_source
    return mod_loader.load_module()
  File "<frozen importlib._bootstrap_external>", line 548, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 1063, in load_module
  File "<frozen importlib._bootstrap_external>", line 888, in load_module
  File "<frozen importlib._bootstrap>", line 290, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 719, in _load
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 879, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1017, in get_code
  File "<frozen importlib._bootstrap_external>", line 947, in source_to_code
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py", line 45
    lib_ext = "*.dll" if compat.is_win else lib_ext = "*.so"
    ^^^^^^^^^^^^^^^^^
SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='?

```

**Comment by rokm** [comment]
> I am getting a SyntaxError:

Yeah, that was a faulty commit. Should be fixed in the latest version now.

**Comment by danyeaw** [comment]
Yup, that is resolved now. Here is the new one:

```
11814 INFO: Loading module hook 'hook-gi.repository.GtkSource.py' from '/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/hooks'...
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/.venv/bin/pyinstaller", line 8, in <module>
    sys.exit(run())
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/__main__.py", line 178, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/__main__.py", line 59, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 842, in main
    build(specfile, distpath, workpath, clean_build)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 764, in build
    exec(code, spec_namespace)
  File "gaphor.spec", line 24, in <module>
    a = Analysis(
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 319, in __init__
    self.__postinit__()
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/datastruct.py", line 173, in __postinit__
    self.assemble()
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/building/build_main.py", line 487, in assemble
    self.graph.process_post_graph_hooks(self)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/analysis.py", line 326, in process_post_graph_hooks
    module_hook.post_graph(analysis)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/imphook.py", line 402, in post_graph
    self._process_hook_func(analysis)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/depend/imphook.py", line 425, in _process_hook_func
    self._hook_module.hook(hook_api)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.10/site-packages/PyInstaller/hooks/hook-gi.repository.GtkSource.py", line 20, in hook
    binaries, datas, hiddenimports = module_info.get_typelib_data()
AttributeError: 'GiModuleInfo' object has no attribute 'get_typelib_data'. Did you mean: 'collect_typelib_data'?
```

**Comment by rokm** [comment]
> AttributeError: 'GiModuleInfo' object has no attribute 'get_typelib_data'. Did you mean: 'collect_typelib_data'?

Thanks! That should be also fixed now.

**Comment by danyeaw** [comment]
Things are working great in Linux and Windows for me 👍

**Review by bwoodsend — APPROVED** [review]

**Comment by rokm** [comment]
> Things are working great in Linux and Windows for me :+1:

Thanks for testing!

**Pull request merged**
